contracts/utils/EmergencyBrake.sol

```diff
@@ -0,0 +1,125 @@
+}
+
+/// @dev EmergencyBrake allows to plan for and execute isolation transactions that remove access permissions for
+/// a target contract from a series of contacts. In an permissioned environment can be used for pausing components.
+/// All contracts in scope of emergency plans must grant ROOT permissions to EmergencyBrake. To mitigate the risk
+/// of governance capture, EmergencyBrake has very limited functionality, being able only to revoke existing roles
+/// and to restore previously revoked roles. Thus EmergencyBrake cannot grant permissions that weren't there in the 
+/// first place. As an additional safeguard, EmergencyBrake cannot revoke or grant ROOT roles.
+/// In addition, there is a separation of concerns between the planner and the executor accounts, so that both of them
+/// must be compromised simultaneously to execute non-approved emergency plans, and then only creating a denial of service.
+contract EmergencyBrake is AccessControl, IEmergencyBrake {
+    enum State {UNKNOWN, PLANNED, EXECUTED, TERMINATED}
+
+    event Planned(bytes32 indexed txHash, address indexed target, address[] indexed contacts, bytes4[][] permissions);
+    event Cancelled(bytes32 indexed txHash, address indexed target, address[] indexed contacts, bytes4[][] permissions);
+    event Executed(bytes32 indexed txHash, address indexed target, address[] indexed contacts, bytes4[][] permissions);
+    event Restored(bytes32 indexed txHash, address indexed target, address[] indexed contacts, bytes4[][] permissions);
+    event Terminated(bytes32 indexed txHash, address indexed target, address[] indexed contacts, bytes4[][] permissions);
+
+    mapping (bytes32 => State) public plans;
+
+    constructor(address planner, address executor) AccessControl() {
+        _grantRole(IEmergencyBrake.plan.selector, planner);
+        _grantRole(IEmergencyBrake.cancel.selector, planner);
+        _grantRole(IEmergencyBrake.execute.selector, executor);
+        _grantRole(IEmergencyBrake.restore.selector, planner);
+        _grantRole(IEmergencyBrake.terminate.selector, planner);
+
+        // Granting roles (plan, cancel, execute, restore, terminate) is reserved to ROOT
+    }
+
+    /// @dev Register an isolation transaction
+    function plan(address target, address[] memory contacts, bytes4[][] memory permissions)
+        external override auth
+        returns (bytes32 txHash)
+    {
+        require(contacts.length == permissions.length, "Mismatched inputs");
+        // Removing or granting ROOT permissions is out of bounds for EmergencyBrake
+        for (uint256 i = 0; i < contacts.length; i++){
+            for (uint256 j = 0; j < permissions[i].length; j++){
+                require(
+                    permissions[i][j] != ROOT,
+                    "Can't remove ROOT"
+                );
+            }
+        }
+        txHash = keccak256(abi.encode(target, contacts, permissions));
```

Comment:
This is wild. I expected you to store the plans on chain, but using the txHash is a very clever way to do this.

Reply:
I got that trick from Compound, it's really smart.